llm/memory.go

```diff
@@ -123,17 +123,22 @@ func EstimateGPULayers(gpus []discover.GpuInfo, ggml *GGML, projectors []string,
 		slog.Warn("model missing blk.0 layer size")
 	}
 
-	kv, graphPartialOffload, graphFullOffload := ggml.GraphSize(uint64(opts.NumCtx), uint64(min(opts.NumCtx, opts.NumBatch)))
+	// Estimate the memory required for KV cache quantization
+	kv := estimateKvCacheSize(kVCacheQuantization(envconfig.KvCacheType(), ggml), uint64(opts.NumCtx), ggml)
```

Comment:
One (larger) issue I had missed until now is that we are no longer using the kv calculations in `ggml.GraphSize`. This is important as the kv size may change based on the architecture (e.g. `mllama`), and that code to calculate `kv` is now unused

I think there's a relatively simple answer in the likes of:
1. Change `estimateKvCacheSize` to a function that returns a bytesize for a given kvCacheType `kvCacheBytesPerElement(cacheType string)`
2. Use `kvCacheBytesPerElement()` instead of the hardcoded value of `2` in `ggml.GraphSize`
3. Change `ggmlGraphSize` to accept a `kvCacheType string` so it can estimate correctly.

In `memory.go` (here), we could do similar to:

```
fa := envconfig.FlashAttention() &&
    gpus.FlashAttentionSupported() &&
    ggml.SupportsFlashAttention()

var kvct string
if fa {
    requested := envconfig.KvCacheType()
    if ggml.SupportsKVCacheType(requested) {
        kvct = requested
    }
}

kv, partialOffload, fullOffload := ggml.GraphSize(context, batch, kvct)
```

and in `ggml.go`:

```
func (ggml GGML) GraphSize(context, batch uint64, kvCacheType string) ... {

// default to 2 bytes per cache element (f16) unless a different
// kvCacheType has been requested
bytesPerCacheElement := 2
if kvCacheType != "" {
    bytesPerCacheElement = kvCacheTypeBytesPerElement(kvCacheType)
}

use `bytesPerCacheElement` instead of `2` below...
```

Note: this change would make testing the full estimation a bit harder. I'm okay with us keeping the tests focused on the new functions (e.g. byteSize checks, support checks etc) vs estimation tests for this PR

Reply:
OK, I think I've actioned each of your points, in the interest of not going insane - as you suggested I've greatly simplified the tests to focus just on the k/v cache quantisation functionality rather than precise estimations.

This is the commit with the changes after your comments: https://github.com/ollama/ollama/pull/6279/commits/4cf0ab75371ef2b8930c42a123a872bc897da15e

I've got to go out now and I'm not sure of my availability over the rest of the weekend, but if there's anything small from you that I notice I'll try to add it in.

Reply:
This is a good catch that we might have architecture specific estimates for KV size. There is one that I see - for mllama we have `2* // sizeof(float16)` that presumably should now be `bytesPerElement`